File: src/utils/diagnosisNormalize.js
/** 将题组扁平化为有序题目列表 */
export function flattenReadingQuestions(groups) {
  const list = []
  for (const g of groups || []) {
    for (const q of g.questions || []) {
      const num = String(q.num ?? q.id?.replace(/^q/i, '') ?? '')
      list.push({
        id: q.id,
        num,
        groupTitle: g.title,
      })
    }
  }
  return sortByNum(list)
}

/** 阅读：从答案键构建全部作答 */
export function buildReadingAllAnswers(userAnswers, answerKey) {
  const all = {}
  for (const [qid, expected] of Object.entries(answerKey || {})) {
    const user = (userAnswers[qid] || '').trim()
    all[qid] = {
      user_ans: user || '—',
      correct_ans: String(expected ?? ''),
    }
  }
  return all
}

/** 听力：从 iframe 批改结果构建题目元数据 */
export function buildListeningQuestionMeta(allAnswers) {
  return sortByNum(
    Object.keys(allAnswers || {}).map((qid) => ({
      id: qid,
      num: String(qid).replace(/^q/i, ''),
    })),
  )
}

function sortByNum(list) {
  return [...list].sort((a, b) => {
    const na = parseInt(a.num, 10)
    const nb = parseInt(b.num, 10)
    if (!Number.isNaN(na) && !Number.isNaN(nb)) return na - nb
    return String(a.num).localeCompare(String(b.num), undefined, { numeric: true })
  })
}

function resolveApiItem(apiItems, q) {
  const items = apiItems || []
  return (
    items.find((i) => i.question_id === q.id) ||
    items.find((i) => i.question_id === q.num) ||
    items.find((i) => i.question_id === `q${q.num}`) ||
    items.find((i) => String(i.label).replace(/^Q/i, '') === String(q.num))
  )
}

function defaultAnalysis(isCorrect, user, correct, mode) {
  if (isCorrect) {
    return mode === 'listening'
      ? '答案正确，注意同义替换与拼写稳定性。'
      : '定位准确，同义替换识别到位。'
  }
  return `错因定位：你将「${user || '未作答'}」误判为正确，正确答案为「${correct}」。`
}

/** 将 API 返回与本地全量题目合并，保证题号 Tab 与刷题区一致 */
export function normalizeDiagnosisItems(apiItems, questionMeta, allAnswers, mode = 'reading') {
  const meta = questionMeta?.length
    ? questionMeta
    : sortByNum(
        Object.keys(allAnswers || {}).map((id) => ({
          id,
          num: String(id).replace(/^q/i, ''),
        })),
      )

  return meta.map((q) => {
    const ans = allAnswers[q.id] || allAnswers[`q${q.num}`] || {}
    const user = (ans.user_ans ?? '').trim()
    const correct = (ans.correct_ans ?? '').trim()
    const localCorrect =
      user &&
      user !== '—' &&
      user.toLowerCase() === correct.toLowerCase()
    const api = resolveApiItem(apiItems, q)

    const isCorrect = api?.is_correct ?? localCorrect

    return {
      question_id: q.id,
      display_num: q.num,
      label: String(q.num),
      is_correct: Boolean(isCorrect),
      correct_answer: api?.correct_answer ?? correct,
      error_analysis: api?.error_analysis ?? defaultAnalysis(isCorrect, user, correct, mode),
      knowledge_point:
        api?.knowledge_point ??
        (mode === 'listening'
          ? '时间信息「先诱后改」+ 同义替换定位'
          : '转折词后信息才是答案 · 词义偷换类干扰'),
      listening_tip:
        api?.listening_tip ??
        (isCorrect
          ? '保持当前精听节奏，重点跟读答案句。'
          : '建议 1.0x 精听转折词前后 2 句，并用 1.25x 跟读答案句。'),
      reading_tip:
        api?.reading_tip ??
        (isCorrect
          ? '保持当前定位节奏，继续积累同义替换。'
          : '建议回读错题所在段落，标出转折词与定位词。'),
      paraphrase_pairs: api?.paraphrase_pairs ?? [],
    }
  })
}

File: backend/app/diagnosis_normalize.py
"""听/读诊断 LLM 输出归一化 — 动态题号字典 → items 数组 + 客观对错强制对齐"""
from __future__ import annotations

from app.diagnosis_items import display_num_for, ordered_question_ids
from app.grading_rules import objective_is_correct


def _normalize_answer(ans) -> tuple[str, str]:
    user = getattr(ans, "user_ans", None) or ans.get("user_ans", "")
    correct = getattr(ans, "correct_ans", None) or ans.get("correct_ans", "")
    return user, correct


def _parse_paraphrase_pairs(raw) -> list[dict]:
    if isinstance(raw, list):
        pairs = []
        for p in raw:
            if isinstance(p, dict):
                pairs.append(
                    {
                        "original": p.get("original", p.get("原文", "")),
                        "replacement": p.get("replacement", p.get("替换", p.get("upgrade", ""))),
                    }
                )
        return [p for p in pairs if p["original"] or p["replacement"]]
    if isinstance(raw, str) and raw.strip():
        pairs = []
        for line in raw.replace("；", "\n").split("\n"):
            line = line.strip()
            if not line:
                continue
            for sep in ("→", "->", "—", "-", "：", ":"):
                if sep in line:
                    a, b = line.split(sep, 1)
                    pairs.append({"original": a.strip(), "replacement": b.strip()})
                    break
            else:
                pairs.append({"original": line, "replacement": ""})
        return pairs
    return []


def _item_from_slice(num: str, slice_data: dict, qid: str, user: str, correct: str, mode: str) -> dict:
    is_correct = objective_is_correct(user, correct)
    status = (slice_data.get("status") or "").strip()
    if status in ("正确", "correct", "true"):
        pass  # still trust objective
    elif status in ("错误", "incorrect", "false"):
        pass

    analysis = (
        slice_data.get("analysis")
        or slice_data.get("error_analysis")
        or ("定位准确，同义替换识别到位。" if is_correct else f"错选「{user}」，正确答案为「{correct}」。")
    )
    suggestion = slice_data.get("suggestion") or slice_data.get("listening_tip") or slice_data.get("reading_tip") or ""
    paraphrase = slice_data.get("paraphrase") or slice_data.get("knowledge_point") or ""

    item = {
        "question_id": qid,
        "display_num": num,
        "label": num,
        "is_correct": is_correct,
        "correct_answer": slice_data.get("correct_answer") or correct,
        "error_analysis": analysis,
        "knowledge_point": paraphrase if isinstance(paraphrase, str) else "",
    }
    if mode == "listening":
        item["listening_tip"] = suggestion or ("保持精听节奏。" if is_correct else "1.0x 精听答案句前后 2 句，标注转折词。")
    else:
        item["reading_tip"] = suggestion or ("回读定位段落，标出转折与同义替换。" if not is_correct else "保持当前定位节奏。")
        item["paraphrase_pairs"] = _parse_paraphrase_pairs(
            slice_data.get("paraphrase_pairs") or paraphrase
        )
    return item


def normalize_diagnosis_result(
    result: dict,
    all_answers: dict,
    question_meta: list | None,
    *,
    mode: str,
) -> list[dict]:
    """将 LLM 返回的 items 数组或 {题号: {...}} 字典统一为前端 items 列表"""
    raw_items = result.get("items") or result.get("diagnosis") or result.get("questions") or {}
    by_num: dict[str, dict] = {}

    if isinstance(raw_items, dict):
        by_num = {str(k): v for k, v in raw_items.items() if isinstance(v, dict)}
    elif isinstance(raw_items, list):
        for it in raw_items:
            if not isinstance(it, dict):
                continue
            num = str(it.get("display_num") or it.get("label") or it.get("question_id", "")).lstrip("qQ")
            by_num[num] = it

    merged: list[dict] = []
    for qid in ordered_question_ids(all_answers, question_meta):
        user, correct = _normalize_answer(all_answers.get(qid) or {})
        num = display_num_for(qid, question_meta)
        slice_data = by_num.get(num) or by_num.get(qid) or by_num.get(str(qid)) or {}
        merged.append(_item_from_slice(num, slice_data, qid, user, correct, mode))
    return merged
